Support internal cross references (in same document with anchor)

anchors like <<name>> produce Exceptions

[ERROR] Failed to execute goal org.sahli.asciidoc.confluence.publisher:asciidoc-confluence-publisher-maven-plugin:0.4.0:publish (default-cli) on project XYZ: Publishing to Confluence failed: Could not convert asciidoc pages: String index out of range: -1 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.sahli.asciidoc.confluence.publisher:asciidoc-confluence-publisher-maven-plugin:0.4.0:publish (default-cli) on project XYZ: Publishing to Confluence failed
..
Caused by: java.lang.RuntimeException: Could not convert asciidoc pages
        at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.convert(AsciidocConfluenceConverter.java:91)
        at org.sahli.asciidoc.confluence.publisher.maven.plugin.AsciidocConfluencePublisherMojo.execute(AsciidocConfluencePublisherMojo.java:84)
        ... 21 more
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -1

--- a/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/helpers.rb
+++ b/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/helpers.rb
@@ -155,4 +155,18 @@ module Slim::Helpers
     (str.include? ':') && str =~ UriSniffRx
   end
 
+  # checks if xref is referring to adoc (internal cross references not yet supported)
+  def cross_page_xref? str
+    str.end_with? ".html"
+  end
+
+  # removes leading hash from anchor targets
+  def anchor_name str
+    if str.start_with? "#"
+      str[1..str.length]
+    else
+      str
+    end
+  end
+
 end

--- a/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/inline_anchor.html.slim
+++ b/asciidoc-confluence-publisher-converter/src/main/resources/org/sahli/asciidoc/confluence/publisher/converter/templates/inline_anchor.html.slim
@@ -1,9 +1,18 @@
 - case type
 - when :xref
-  ac:link
-    ri:page ri:content-title=target
-    ac:plain-text-link-body
-      <![CDATA[#{(xref_text || %([#{attr :refid}]))}]]>
+  - if cross_page_xref? target
+    ac:link
+      ri:page ri:content-title=target
+      ac:plain-text-link-body
+        <![CDATA[#{(xref_text || %([#{attr :refid}]))}]]>
+  - else
+    ac:link ac:anchor=(anchor_name @target)
+      ac:plain-text-link-body
+        <![CDATA[#{(xref_text || %([#{attr :refid}]))}]]>
+- when :ref
+  ac:structured-macro ac:name="anchor"
+    ac:parameter ac:name=""
+     =(anchor_name target)
 - when :link
   - if uriish? target
     a href=target

--- a/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
+++ b/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
@@ -974,6 +974,74 @@ public class AsciidocConfluencePageTest {
     }
 
     @Test
+    public void renderConfluencePage_asciiDocWithInternalCrossReferenceToSection_returnsConfluencePageContentWithInternalCrossReferenceToSectionUsingSectionTitle() {
+        // arrange
+        String adocContent = "" +
+                "== Section 1 [[section1]]\n" +
+                "Cross reference to <<section1>>";
+
+        // act
+        AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage(prependTitle(adocContent)), UTF_8, TEMPLATES_FOLDER, dummyAssetsTargetPath());
+
+        // assert
+        String expectedContent = "" +
+                "<h1>Section 1</h1>" +
+                "<p>Cross reference to <ac:link ac:anchor=\"section1\"><ac:plain-text-link-body><![CDATA[Section 1]]></ac:plain-text-link-body></ac:link></p>";
+        assertThat(asciidocConfluencePage.content(), is(expectedContent));
+    }
+
+    @Test
+    public void renderConfluencePage_asciiDocWithInternalCrossReferenceToSectionAndCustomLabel_returnsConfluencePageContentWithInternalCrossReferenceToSectionUsingCustomLabel() {
+        // arrange
+        String adocContent = "" +
+                "== Section 1 [[section1]]\n" +
+                "Cross reference to <<section1,section 1>>";
+
+        // act
+        AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage(prependTitle(adocContent)), UTF_8, TEMPLATES_FOLDER, dummyAssetsTargetPath());
+
+        // assert
+        String expectedContent = "" +
+                "<h1>Section 1</h1>" +
+                "<p>Cross reference to <ac:link ac:anchor=\"section1\"><ac:plain-text-link-body><![CDATA[section 1]]></ac:plain-text-link-body></ac:link></p>";
+        assertThat(asciidocConfluencePage.content(), is(expectedContent));
+    }
+
+    @Test
+    public void renderConfluencePage_asciiDocWithInternalCrossReferenceToParagraph_returnsConfluencePageContentWithInternalCrossReferenceToParagraphUsingAnchorId() {
+        // arrange
+        String adocContent = "" +
+                "[[paragraph1]]Paragraph\n\n" +
+                "Cross reference to <<paragraph1>>";
+
+        // act
+        AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage(prependTitle(adocContent)), UTF_8, TEMPLATES_FOLDER, dummyAssetsTargetPath());
+
+        // assert
+        String expectedContent = "" +
+                "<p><ac:structured-macro ac:name=\"anchor\"><ac:parameter ac:name=\"\">paragraph1</ac:parameter></ac:structured-macro>Paragraph</p>\n" +
+                "<p>Cross reference to <ac:link ac:anchor=\"paragraph1\"><ac:plain-text-link-body><![CDATA[[paragraph1]]]></ac:plain-text-link-body></ac:link></p>";
+        assertThat(asciidocConfluencePage.content(), is(expectedContent));
+    }
+
+    @Test
+    public void renderConfluencePage_asciiDocWithInternalCrossReferenceToParagraphAndCustomLabel_returnsConfluencePageContentWithInternalCrossReferenceToParagraphUsingCustomLabel() {
+        // arrange
+        String adocContent = "" +
+                "[[paragraph1]]Paragraph\n\n" +
+                "Cross reference to <<paragraph1,Paragraph>>";
+
+        // act
+        AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage(prependTitle(adocContent)), UTF_8, TEMPLATES_FOLDER, dummyAssetsTargetPath());
+
+        // assert
+        String expectedContent = "" +
+                "<p><ac:structured-macro ac:name=\"anchor\"><ac:parameter ac:name=\"\">paragraph1</ac:parameter></ac:structured-macro>Paragraph</p>\n" +
+                "<p>Cross reference to <ac:link ac:anchor=\"paragraph1\"><ac:plain-text-link-body><![CDATA[Paragraph]]></ac:plain-text-link-body></ac:link></p>";
+        assertThat(asciidocConfluencePage.content(), is(expectedContent));
+    }
+
+    @Test
     public void attachments_asciiDocWithImage_returnsImageAsAttachmentWithPathAndName() {
         // arrange
         String adocContent = "image::sunset.jpg[]";

--- a/asciidoc-confluence-publisher-doc/etc/docs/00-index/01-pages.adoc
+++ b/asciidoc-confluence-publisher-doc/etc/docs/00-index/01-pages.adoc
@@ -11,6 +11,7 @@ must contain a top-level document title which will be used as the Confluence pag
 According to the restrictions of Confluence, document titles must be unique within the Confluence space the
 documentation is published to. The AsciiDoc file name itself is not reflected in Confluence.
 
+
 == Links between Pages
 
 Links between pages in Confluence are supported using AsciiDoc inter-document cross references. Target pages are
@@ -26,7 +27,24 @@ referencing AsciiDoc file. All referenced AsciiDoc files must be placed below th
 <<01-pages.adoc#, Link to Pages>>
 <<../00-index/06-images.adoc#, Link to Images>>
 
+
+== Links within Pages
+
+Links within pages in Confluence are supported using AsciiDoc inline anchors and internal cross references:
+
+[listing]
+....
+[[paragraph-a]]Paragraph with inline anchor
+
+Another paragraph with internal cross-reference to <<paragraph-a>>
+....
+
+[[paragraph-a]]Paragraph with inline anchor
+
+Another paragraph with internal cross-reference to <<paragraph-a>>
+
+
 [WARNING]
 ====
-Links within the same page are currently not supported.
+Links from one page to an inline anchor within another page are currently not supported.
 ====

